tests/kernel/mem_protect/userspace/src/main.c

```diff
@@ -1151,6 +1151,45 @@ void test_oops_stackcheck(void)
 	test_oops(K_ERR_STACK_CHK_FAIL, K_ERR_STACK_CHK_FAIL);
 }
 
+struct k_timer user_timer;
+ZTEST_BMEM volatile bool timer_fired;
+
+void user_timer_func(struct k_timer *unused)
+{
+	ARG_UNUSED(unused);
+
+	printk("timer fired\n");
+	timer_fired = true;
+}
+
+/* Test scenario where a user thread is running and gets preempted by a timer
+ * interrupt while still in user mode (and not in, for example, a system call).
+ * When we return, verify that we are still in user mode.
+ */
+void test_survive_interrupt(void)
+{
+	k_timer_start(&user_timer, 200, 200);
+	printk("spinning\n");
+	while (timer_fired == false) {
+		/* spin */
+	}
+	printk("done\n");
+	k_timer_stop(&user_timer);
+
+	timer_fired = false;
+
+	k_timer_start(&user_timer, 200, 200);
+	printk("spinning\n");
+	while (timer_fired == false) {
+		/* spin */
+	}
+	printk("done\n");
+	k_timer_stop(&user_timer);
+
```

Comment:
It seems that line 1171-1177 is same as line 1181-1187. Can we put them in a "for( ; ; ) { }"?